src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs

```diff
@@ -74,10 +74,12 @@ public static async ValueTask<SslStream> EstablishSslConnectionAsync(SslClientAu
                         sslStream.AuthenticateAsClient(sslOptions);
                     }
                 }
+                activity?.AddEvent(new ActivityEvent("Tls.Authenticated"));
```

Comment:
I'd prefer to use spans and avoid using span events at all costs:
1. A child span for TLS handshake would have a duration and timestamp
2. It won't be verbose since connections are long-lived (much longer than requests)
3. It can live without HTTP instrumentation on
4. It can be turned off (if coming from a different source) while HTTP is on
5. Span events are likely to be deprecated in the future in favor of independent Event API similar to logging
6. We'll have standard means to record errors/duration.

Reply:
> I'd prefer to use spans and avoid using span events at all costs:

That seems fine to me. I mentioned in another comment that there was a timing issue where one long activity for the entire connection wouldn't transmit the events until the connection closed. I suggested changing the events to log messages as one potential solution, but changing them to spans also solves that problem.

> It won't be verbose since connections are long-lived (much longer than requests)

I assume at the limit we might have one connection per request, but I hope such scenarios are uncommon in any high throughput service. Worst case if connection activity overhead was too high the app developer can always sample them or turn them off.

Reply:
I like the advantages that come from implementing this with sub activities. Another big advantage is that it would also work if `ConnectCallback` is present, since we would implement them in downstream libraries. However, there are some challanges:
- It increases the scope of the problem, since we would need to touch multiple `System.Net.*` libraries. We may need to pull in more reviewers from the networking team.
- If we want to add attributes to those sub-activities, it increases the scope even further since we would need to make decisions on the characteristics of those attributes (eg. naming, possible values for attributes like `error.type`, etc). As I also noted in https://github.com/dotnet/runtime/pull/101814#discussion_r1595495720, my recommendation would be to track this in a separate issue.
-  If there is an `Activity.Current` to be used as a parent activity, and `System.Net.Http.Connections.HttpConnectionOut` is not enabled, the DNS/TCP/TLS activity would attach to the Http activity as parent which is somewhat unfortunate. We would need to document that users should avoid this.

Regardless, I agree sub-activities are better. If we have consensus, I will change this PR to add activities in downstream libs. @MihaZupan any concerns?

>  changing the events to log messages

Just to not exclude this option completely yet: @noahfalk do you mean a writing to `DiagnosticListener` by this or some other technique?

Reply:
> If there is an Activity.Current to be used as a parent activity, and System.Net.Http.Connections.HttpConnectionOut is not enabled, the DNS/TCP/TLS activity would attach to the Http activity as parent which is somewhat unfortunate. We would need to document that users should avoid this.

This is why when we originally talked I said not to use Activity.Current and be explicit about passing in the activity if you can. 
If the DNS and/or TLS activities are invoked and the parent is not a connection (As the source isn't enabled) , then they can look at the activity and see if its a connection or not and therefore not parent themselves to it. You will still see them on the overall timeline.

The other hack would be to put all these activities in the same source, so they can only be enabled/disabled as a group. But as I think they would be useful for other scenarios beyond http connections, that really does seem like a hack.

Reply:
I was thinking about something that would ultimately be ILogger but later recalled that I don't think ILogger is easy to generate from the level of this code in the stack. Technically there are ways around that but I wouldn't worry about pursuing it any further. I think sub-activities are both easier and better.

Reply:
> But as I think they would be useful for other scenarios beyond http connections

This is why I would prefer to create the sub-activities in downlevel libs. `System.Net.Sockets|NameResolution|Security` would all have their own ActivitySource-s.

> passing in the activity if you can.

It's not possible to pass it to downlevel libraries without significant (and IMO bloated) API additions, which we likely don't want.